best-practices: add massive region performance tuning best practices (#1704)
pingcap/docs: dev/reference/best-practices/massive-regions.md

## Performance problem

From the diagram of Raftstore workflow, messages for each Region are processed one by one. When Regions are in huge numbers, it takes Raftstore sometime to process the heartbeats of these Regions, which can cause some delay. As a result, some read and write requests are not processed in time. If read and write pressure is high, CPU usage of the Raftstore thread might easily reach the bottleneck, which further increases the delay and affects the performance.

Generally, if the CPU usage of the loaded Raftstore reaches 85% or higher, Raftstore is in busy state and becomes the bottleneck. At the same time, `propose wait duration` can be as high as hundreds of milliseconds.

> **Note:**
>

> > + For the CPU usage of Raftstore as mentioned above, Raftstore is single-threaded. If Raftstore is multi-threaded, CPU usage threshold (85%) can be increased proportionally.

Reviewer: ```suggestion
> + For the CPU usage of Raftstore as mentioned above, Raftstore is single-threaded. If Raftstore is multi-threaded, you can increase the CPU usage threshold (85%) proportionally.
```

> + Because I/O operations exist in the Raftstore thread, CPU usage cannot reach 100%.

### Performance monitoring

> You can check the following monitoring metrics in Grafana's *TiKV* panel:

Reviewer: ```suggestion
You can check the following monitoring metrics in Grafana's **TiKV Dashboard**:
```

> + `Raft store CPU` in the *Thread-CPU* panel

Reviewer: Generally, use the bold style for words in the UI interface. Please update similar occurrences in this PR. Thanks!

### Method 3: Increase Raftstore concurrency

Raftstore in TiDB v3.0 has been upgraded to a multi-threaded module, which greatly reduces the possibility that a Raftstore thread becomes the bottleneck.

By default, `raftstore.store-pool-size` is configured to `2` in TiKV. If a bottleneck occurs in Raftstore, you can properly increase the value of this configuration according to actual situation. But to avoid introducing unnecessary thread switching overhead, it is recommended that you do not set this value too high.

### Method 4: Enable Hibernate Region

In the actual situation, read and write requests are not evenly distributed on every Region. Instead, they are concentrated on a few Regions. Then you can minimize the number of messages for the temporarily idle Regions, which is the feature of Hibernate Region. In this feature, Raftstore does sent tick messages to the Raft state machines of idle Regions when necessary. Then these Raft state machines will not be triggered to generate heartbeat information, greatly reducing the workload of Raftstore.

> Up to TiDB v3.0.5, Hibernate Region is still an experimental feature, which is enabled by default in [TiKV master](https://github.com/tikv/tikv/tree/master). You can enable this feature according to your actual need. For the configuration of Hibernate Region, refer to [Configure Hibernate Region](https://github.com/tikv/tikv/blob/master/docs/reference/configuration/raftstore-config.md#hibernate-region).

Reviewer: ```suggestion
Up to TiDB v3.0.5, Hibernate Region is still an experimental feature, which is enabled by default in [TiKV master](https://github.com/tikv/tikv/tree/master). You can enable this feature according to your needs. For the configuration of Hibernate Region, refer to [Configure Hibernate Region](/dev/reference/configuration/tikv-server/configuration-file.md#raftstorehibernate-regions-experimental).
```

### Method 5: Enable `Region Merge`

> **Note:**
>
> `Region Merge` is enabled in TiDB v3.0 by default.

You can also reduce the number of Regions by enabling `Region Merge`. Contrary to `Region Split`, `Region Merge` is the process of merging adjacent small Regions through scheduling. After dropping data or executing the `Drop Table` or `Truncate Table` statement, you can merge small Regions or even empty Regions to reduce resource consumption.

> Enable `Region Merge` by setting the following configurations:

Reviewer: ```suggestion
Enable `Region Merge` by configuring the following parameters:
```

{{< copyable "" >}}

```
>> pd-ctl config set max-merge-region-size 20
>> pd-ctl config set max-merge-region-keys 200000
>> pd-ctl config set merge-schedule-limit 8
```

> Refer to [Region Merge](https://github.com/tikv/tikv/blob/master/docs/how-to/configure/region-merge.md) and [PD configuration file](/dev/reference/configuration/pd-server/configuration-file.md#schedule) for more details.

Reviewer: ```suggestion
Refer to [Region Merge](https://github.com/tikv/tikv/blob/master/docs/how-to/configure/region-merge.md) and the following three configuration parameters in the [PD configuration file](/dev/reference/configuration/pd-server/configuration-file.md#schedule) for more details:

- [`max-merge-region-size`](/dev/reference/configuration/pd-server/configuration-file.md#max-merge-region-size)
- [`max-merge-region-keys`](/dev/reference/configuration/pd-server/configuration-file.md#max-merge-region-keys)
- [`merge-schedule-limit`](/dev/reference/configuration/pd-server/configuration-file.md#merge-schedule-limit)
```

The default configuration of the `Region Merge` parameters is rather conservative. You can speed up the `Region Merge` process by referring to the method provided in [PD Scheduling Best Practices](/dev/reference/best-practices/pd-scheduling.md#region-merge-is-slow).

## Other problems and solutions

> 

Reviewer: ```suggestion

This section describes some other problems and solutions.

```

### Switching PD Leader is slow

> PD needs to persist Region Meta information on etcd to ensure that PD can quickly resume to provide Region routing services after switching the PD Leader node. As the number of Regions increases, the performance problem of etcd appears, making it slower for PD to get Region Meta information from etcd when PD is switching the Leader. With millions of Regions, it might take dozens or even tens of seconds to get the meta information from etcd.

Reviewer: ```suggestion
PD needs to persist Region Meta information on etcd to ensure that PD can quickly resume to provide Region routing services after switching the PD Leader node. As the number of Regions increases, the performance problem of etcd appears, making it slower for PD to get Region Meta information from etcd when PD is switching the Leader. With millions of Regions, it might take more than ten seconds or even tens of seconds to get the meta information from etcd.
```

> To address this problem, `use-region-storage` is enabled by default in PD since TiDB v3.0. With this feature enabled, PD stores Region Meta information on local LevelDB and synchronizes the information among PD nodes through other mechanism.

Reviewer: Please confirm whether the **"synchronize"** is correctly used here. Is it just "obtain" or another meaning?

```suggestion
To address this problem, `use-region-storage` is enabled by default in PD in TiDB v3.0. With this feature enabled, PD stores Region Meta information on local LevelDB and synchronizes the information among PD nodes through other mechanisms.
```

Author: **"synchronize"** is correct here, as confirmed with the author.

### PD routing information is not updated in time

> In TiKV, pd-worker regularly reports Region Meta information to PD. When TiKV is restarted or switches the Region Leader, PD needs to recalculate Region's `approximate size / keys` through statistics. Therefore, with a large number of Regions, the single-threaded pd-worker might become the bottleneck, causing tasks to be piled up and not processed in time. In this situation, PD cannot obtain certain Region Meta information in time so that the routing information is not updated in time. This problem does not affect the actual reads and writes, but might cause inaccurate PD scheduling and require several round trips when TiDB updates Region cache.

Reviewer: ```suggestion
In TiKV, pd-worker regularly reports Region Meta information to PD. When TiKV is restarted or switches the Region leader, PD needs to recalculate Region's `approximate size / keys` through statistics. Therefore, with a large number of Regions, the single-threaded pd-worker might become the bottleneck, causing tasks to be piled up and not processed in time. In this situation, PD cannot obtain certain Region Meta information in time so that the routing information is not updated in time. This problem does not affect the actual reads and writes, but might cause inaccurate PD scheduling and require several round trips when TiDB updates Region cache.
```

> You can check *Worker pending tasks* under *Task* in the *TiKV Grafana* panel to determine whether pd-worker has tasks piled up. Generally, `pending tasks` should be kept at a relatively low value.

Reviewer: ditto. UI words format.

![Check pd-worker](/media/best-practices/pd-worker-metrics.png)

> Currently, pd-worker is optimized for better efficiency in [#5620](https://github.com/tikv/tikv/pull/5620) on [TiKV master](https://github.com/tikv/tikv/tree/master), which is applied since [v3.0.5](/dev/releases/3.0.5.md#tikv). If you encounter a similar problem, it is recommended to upgrade to v3.0.5 or later version.

Reviewer: ```suggestion
Currently, pd-worker is optimized for better efficiency in [#5620](https://github.com/tikv/tikv/pull/5620) on [TiKV master](https://github.com/tikv/tikv/tree/master), which is applied since [v3.0.5](/dev/releases/3.0.5.md#tikv). If you encounter a similar problem, it is recommended to upgrade to v3.0.5 or later versions.
```

### Prometheus is slow to query metrics

> In a large-scale cluster, as the number of TiKV instances increases, Prometheus has greater pressure to query metrics, making it slower for Grafana to display these metrics. To ease this problem, pre-calculations of metrics is configured in v3.0.

Reviewer: ```suggestion
In a large-scale cluster, as the number of TiKV instances increases, Prometheus has greater pressure to query metrics, making it slower for Grafana to display these metrics. To ease this problem, metrics pre-calculation is configured in v3.0.
```
